compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala

```diff
@@ -0,0 +1,426 @@
+    if (ctx.compilationUnit.needsStaging) super.run(freshStagingContext)
+
+  protected def newTransformer(implicit ctx: Context): Transformer = new Transformer {
+    override def transform(tree: tpd.Tree)(implicit ctx: Context): tpd.Tree =
+      new QuoteReifier(null, new mutable.HashMap[Symbol, Tree => Tree], new Embedded, ctx.owner)(ctx).transform(tree)
+  }
+
+  /** The main transformer class
+   *  @param  outer      the next outer reifier, null is this is the topmost transformer
+   *  @param  embedded   a list of embedded quotes (if in a splice) or splices (if in a quote)
+   *  @param  owner      the owner in the destination lifted lambda
+   *  @param  capturers  register a reference defined in a quote but used in another quote nested in a splice.
+   *                     Returns a version of the reference that needs to be used in its place.
+   *                     '{
+   *                       val x = ???
+   *                       { ... '{ ... x ... } ... }.unary_~
+   *                     }
+   *                     Eta expanding the `x` in `{ ... '{ ... x ... } ... }.unary_~` will return a `x$1.unary_~` for which the `x$1`
+   *                     be created by some outer reifier.
+   *                     This transformation is only applied to definitions at staging level 1.
+   *                     See `isCaptured`.
+   */
+  private class QuoteReifier(outer: QuoteReifier, capturers: mutable.HashMap[Symbol, Tree => Tree],
+                             val embedded: Embedded, val owner: Symbol)(@constructorOnly ictx: Context) extends TreeMapWithStages(ictx) { self =>
+
+    import TreeMapWithStages._
+
+    /** A nested reifier for a quote (if `isQuote = true`) or a splice (if not) */
+    def nested(isQuote: Boolean)(implicit ctx: Context): QuoteReifier = {
+      val nestedEmbedded = if (level > 1 || (level == 1 && isQuote)) embedded else new Embedded
+      new QuoteReifier(this, capturers, nestedEmbedded, ctx.owner)(ctx)
+    }
+
+    /** Assuming <expr> contains types `<tag1>.unary_~, ..., <tagN>.unary_~`, the expression
+     *
+     *      { type <Type1> = <tag1>.unary_~
+     *        ...
+     *        type <TypeN> = <tagN>.unary_~
+     *        <expr>
+     *      }
+     *
+     *  references to `TypeI` in `expr` are rewired to point to the locally
+     *  defined versions. As a side effect, prepend the expressions `tag1, ..., `tagN`
+     *  as splices.
+     */
+    private def addTags(expr: Tree)(implicit ctx: Context): Tree = {
+
+      def mkTagSymbolAndAssignType(spliced: TermRef): TypeDef = {
+        val splicedTree = tpd.ref(spliced)
+        val rhs = transform(splicedTree.select(tpnme.UNARY_~))
+        val alias = ctx.typeAssigner.assignType(untpd.TypeBoundsTree(rhs, rhs), rhs, rhs)
+        val local = ctx.newSymbol(
+          owner = ctx.owner,
+          name = UniqueName.fresh((splicedTree.symbol.name.toString + "$" + nme.UNARY_~).toTermName).toTypeName,
```

Comment:
Why is it `<name>$unary_~` and not simply `<name>$_~`?

Reply:
Good point. Will change it.